requirements/specifications/general/context-parameters.md

```diff
@@ -0,0 +1,303 @@
+    - [4.1.4. Event RPC Decoration](#414-event-rpc-decoration)
+  - [4.2. JavaScript API](#42-javascript-api)
+    - [4.2.1. Event Listener Signatures](#421-event-listener-signatures)
+
+## 3. Context Parameters Use Cases
+
+How each parameter affects it's corresponding API is out of scope for
+this document. See each API spec for details on what each context
+parameter does.
+
+Setting a context parameter to null is still setting it to a value. If a
+context parameter is passed to the SDK with either a value of null or
+undefined, then it **MUST** be explicitly set in the RPC request to the
+value null. This is to avoid additional method signature permutations
+being required for the SDK.
+
+An effort should be made to sort the context parameters in order of most
+usefulness, since not all languages support undefined.
+
+## 4. Context Parameters API
+
+The section describes the RPC and JavaScript APIs.
+
+### 4.1. JSON-RPC API
+
+Simple getters and event listeners don't need any parameters.
+
+If a method is tagged as either a property (any kind) or an event, then
+**all** the parameters in the RPC definition **MUST** be context
+parameters.
+
+To facilitate this, the listen parameter that all events currently have
+will be removed from the source module and inserted into the generated
+RPC by the firebolt-openrpc tooling.
+
+If any Context Parameters have a type other than:
+
+-   `string`
+-   `boolean`
+-   `number`
+-   `integer`
+
+Then the RPC method **MUST NOT** pass validation. This is to ensure that
+implementing context parameters is not overly complicated.
+
+#### 4.1.1. Setter RPC generation
+
+When generating the setter for a property method, all the Context
+Parameters **MUST** be copied to the setter. The context parameters
+**MUST** be before the value parameter, which itself **MUST** be last.
+
+#### 4.1.2. on<Property>Changed RPC generation
+
+When generating the on<Property>Changed notification for a property
+method, all the Context Parameters **MUST** be copied to the event
+parameters. The context parameters **MUST** be before the listen
+parameter, which itself **MUST** be last.
+
+#### 4.1.3. Temporal Set on<Item>Available / Unavailable RPC generation
+
+Generated Temporal Set events will treat the entire parameter list from
+the original temporal-set method as Context Parameters and copy them to
+the event parameters. The context parameters **MUST** be before the
+listen parameter, which itself **MUST** be last.
+
+#### 4.1.4. Event RPC Decoration
+
+All RPC methods tagged as event **MUST** have the listen parameter, of
+type boolean, automatically added when generating the final RPC. This
+parameter will be added at the end of the parameters list.
+
+All RPC methods tagged as event that have context parameters **MUST**
+have the result schema wrapped in an object. The original result schema
+**MUST** be moved to a property called data. Each of the context
+parameters **MUST** be copied to a property called context.
```

Comment:
Do we do this for onChange RPC events for properties, otherwise why are they different?

Reply:
Yes, onFooChanged events get the same treatment.